hoglet.go

```diff
@@ -118,19 +135,31 @@ func (c *Circuit[IN, OUT]) setOpenedAt(i int64) {
 // Call calls the wrapped function if the circuit is closed and returns its result. If the circuit is open, it returns
 // [ErrCircuitOpen].
 //
-// The wrapped function is called synchronously, but possilble context errors are recorded as soon as they occur. This
+// The wrapped function is called synchronously, but possible context errors are recorded as soon as they occur. This
 // ensures the circuit opens quickly, even if the wrapped function blocks.
 //
 // By default, all errors are considered failures (including [context.Canceled]), but this can be customized via
 // [WithFailureCondition] and [IgnoreContextCancelation].
 //
 // Panics are observed as failures, but are not recovered (i.e.: they are "repanicked" instead).
 func (c *Circuit[IN, OUT]) Call(ctx context.Context, in IN) (out OUT, err error) {
-	obs := c.observerForCall()
-	if obs == nil {
-		return out, ErrCircuitOpen
+	if c.f == nil {
+		return out, nil
+	}
+
+	obs, err := c.observerForCall(ctx)
+	if err != nil {
+		// Note: errors here are not "observed" and do not count towards the breaker's failure rate.
+		// This includes:
+		// - ErrCircuitOpen (so no "feedback loops")
+		// - ErrConcurrencyLimit (for blocking limited circuits)
+		// - context timeouts while blocked on concurrency limit
+		return out, err
 	}
 
+	// ensure we dedup the final - potentially wrapped - observer.
+	obs = dedupObservableCall(obs)
```

Comment:
I don't quite undertand this here, shouldn't we dedup per circuit and not per call? If we call multiple times, we're creating a `sync.Once` per call and thus dedupping only for each call? 🤔 

```
func dedupObservableCall(obs observer) observer {
	o := sync.Once{}
	return observableCall(func(failure bool) {
		o.Do(func() {
			obs.observe(failure)
		})
	})
}
```

Reply:
The idea is deduplicating the observation of a result. This is important because we do async context tracking: if the context times out, we want to notice it quickly, but leave the actual wrapped function run in the same goroutine to avoid all the racyness of having it (maybe) still run in the background after the wrapped call returns.

So to avoid the race between this concurrent context observing and the observing of the actual result of the function, we have the `sync.Once` in the observer itself.

Does that make sense? :sweat_smile:

Reply:
Ah now it get, it's also just for observing (as the name would suggest 🙈), thanks!